### PR TITLE
Validate pseudo and conditional HTTP/2 headers

### PR DESCRIPTION
**Pull request by madgnome**
**Motivation**:

Netty HTTP/2 implementation is not 100% compliant to the spec ( https://github.com/netty/netty/issues/5761 ). This
commit improves the compliance regarding headers validation,
in particular pseudo-headers and connection ones.

According to the spec:

>    All HTTP/2 requests MUST include exactly one valid value for the
>    ":method", ":scheme", and ":path" pseudo-header fields, unless it is
>    a CONNECT request (Section 8.3).  An HTTP request that omits
>    mandatory pseudo-header fields is malformed (Section 8.1.2.6).

**Modifications**:

- Introduce Http2HeadersValidator class capable of validating HTTP/2
headers
- Invoke validation from DefaultHttp2ConnectionDecoder#onHeadersRead
- Modify tests to use valid headers when required
- Modify HttpConversionUtil#toHttp2Headers to not add :scheme and
:path header on CONNECT method in order to conform to the spec

**Result**:

- Initial requests without :method, :path, :scheme will fail
- Initial requests with multiple values for :method, :path, :scheme
will fail
- Initial requests with an empty :path fail
- Requests with connection-specific header field will fail
- Requests with TE header different than "trailers" will fail
- Fixes 8.1.2.2 tests from h2spec #5761
- Fixes 8.1.2.3 tests from h2spec #5761

### CODE REVIEW

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by madgnome** [comment]
@netty-bot test this please

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java [hunk 1]
```diff
@@ -42,7 +42,7 @@ public boolean process(byte value) {
         public void validateName(CharSequence name) {
             if (name == null || name.length() == 0) {
                 PlatformDependent.throwException(connectionError(PROTOCOL_ERROR,
-                        "empty headers are not allowed [%s]", name));
+                        "empty headers are not allowed"));
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java]: why this change ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:scheme] must be omitted when using CONNECT method.");
+            }
+
+            if (headers.contains(PATH.value())) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:path] must be omitted when using CONNECT method.");
+            }
+        } else {
+            for (AsciiString header : mandatoryHeaders) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: This will allocate a new Iterator on each loop. Consider using an old style for loop.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    throw streamError(streamId, PROTOCOL_ERROR,
+                                      "Mandatory header [%s] is missing.", header);
+                } else if (valueSize > 1) {
+                    throw streamError(streamId, PROTOCOL_ERROR,
+                                      "Header [%s] should have a unique value.", header);
+                }
+            }
+
+            final CharSequence path = headers.get(PATH.value());
+            if (path != null && path.length() == 0) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: should we move this check above the for loop as the check is less expensive.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.3">RFC7540, section-8.1.2.3</a>
+     */
+    static void validatePseudoHeaders(Http2Headers headers, int streamId) throws Http2Exception {
+        final CharSequence method = headers.get(METHOD.value());
+        if (method == null) {
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "Mandatory header [:method] is missing.");
+        }
+
+        if (HttpMethod.CONNECT.asciiName().contentEqualsIgnoreCase(method.toString())) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: toString() is not needed here.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2TestUtil.java [hunk 5]
```diff
@@ -80,6 +80,15 @@ public void run() {
         return data;
     }
 
+    public static Http2Headers defaultWithPseudoHeaders() {
+        Http2Headers headers = new DefaultHttp2Headers(true);
+        headers.method("GET");
+        headers.path("/");
+        headers.scheme("https");
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/Http2TestUtil.java]: use method chaining ?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2TestUtil.java [hunk 6]
```diff
@@ -80,6 +80,15 @@ public void run() {
         return data;
     }
 
+    public static Http2Headers defaultWithPseudoHeaders() {
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/Http2TestUtil.java]: should we maybe rename tis this `newHttp2HeadersWithPseudoHeaders()`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java (hunk 1 above)
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java]: Oops will change that back

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.3">RFC7540, section-8.1.2.3</a>
+     */
+    static void validatePseudoHeaders(Http2Headers headers, int streamId) throws Http2Exception {
+        final CharSequence method = headers.get(METHOD.value());
+        if (method == null) {
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "Mandatory header [:method] is missing.");
+        }
+
+        if (HttpMethod.CONNECT.asciiName().contentEqualsIgnoreCase(method)) {
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Do we need to check for multiple `:method` headers?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        } else {
+            final CharSequence path = headers.get(PATH.value());
+            if (path != null && path.length() == 0) {
+                throw streamError(streamId, PROTOCOL_ERROR, "[:path] header cannot be empty.");
+            }
+
+            for (int i = 0; i < mandatoryHeaders.size(); i++) {
+                AsciiString header = mandatoryHeaders.get(i);
+                final int valueSize = headers.getAll(header).size();
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: This is going to be relatively poor performing considering we're going to make a list for each mandatory header. Since the spec clearly defines which ones are mandatory could we just iterate through the header list in order and count them up, eg have `boolean methodObserved, pathObserved, ...` and make sure we don't do more than one and that we got all three?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java [hunk 9]
```diff
@@ -571,6 +570,22 @@ private void verifyStreamMayHaveExisted(int streamId) throws Http2Exception {
                 throw connectionError(PROTOCOL_ERROR, "Stream %d does not exist", streamId);
             }
         }
+
+        private void validateHeaders(int streamId, Http2Headers headers, Http2Stream stream, boolean endOfStream,
+                                     boolean isInformational) throws Http2Exception {
+            if (connection.isServer() &&
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: Do we worry about the client side validation somewhere else?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java (hunk 7 above)
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Ah yeah good point! I should add that to h2spec as well

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java (hunk 8 above)
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Iterating through the whole list would scale linearly with the number of headers whereas this version has a constant run time. If we assume the number of headers is most of the time small iterating would be more performant.

Ideally we would have a `count(name)` method on headers that would just count the headers with a given name, not convinced this usecase justify adding that to the API though.

Comment (bryce-anderson) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: If we use my strategy of having an `observed` variable for each critical header we should count no more than 4 as at 4 we'd hit an exceptional condition. This actually gets to another conformance requirement of the RFC ([section 8.1.2.1](https://http2.github.io/http2-spec/#rfc.section.8.1.2.1)) which is that we should make sure we don't have bonus pseudo headers: `Pseudo-header fields defined for requests MUST NOT appear in responses; pseudo-header fields defined for responses MUST NOT appear in requests.`

Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Nice, forgot about this requirement, will change it. 
(BTW 8.1.2.1 requirement is already checked in the HPackDecoder)

Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: They validate at different levels: in `HPackdecoder` we check if the headerBlock is valid, here we check if all the headers together make sense, that being said it would be nice if it was done at a single place.

The problem is that `HPackDecoder` level validation needs to know the order of the headers, we lose the order once we return Http2Headers and  `Http2HeadersValidation` needs to know more about the context: is this an initial header or trailing one? It didn't seem right to push that down to hpackdecoder.

Comment (bryce-anderson) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: That's true regarding the order validation in hpack. Maybe we could shift request/response validation to where you're adding more validation logic, but probably not all of it and at a minimum this isn't obviously right.

Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: > If we use my strategy of having an observed variable for each critical header we should count no more than 4 as at 4 we'd hit an exceptional condition.

Thinking about it more, we don't have this guarantee since the headers order is not guaranteed in `Http2Headers`.
The implementation seems to keep the insertion order but this is not specified so potentially subject to change later.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java (hunk 9 above)
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: Baby steps, h2spec only validate server specification for now.

**Comment by madgnome** [comment]
I've actioned PR feedbacks:
- provided a way to disable validation
- change the way we validate pseudo headers to be more efficient
- move request/response validation from HPackDecoder to Http2HeadersValidator

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 10]
```diff
@@ -388,13 +388,8 @@ private static HeaderType validate(int streamId, CharSequence name,
...
             }
 
-            final HeaderType currentHeaderType = pseudoHeader.isRequestOnly() ?
+            return pseudoHeader.isRequestOnly() ?
                     HeaderType.REQUEST_PSEUDO_HEADER : HeaderType.RESPONSE_PSEUDO_HEADER;
-            if (previousHeaderType != null && currentHeaderType != previousHeaderType) {
-                throw streamError(streamId, PROTOCOL_ERROR, "Mix of request and response pseudo-headers.");
-            }
-
-            return currentHeaderType;
```
Comment (madgnome) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java]: This check was moved to `Http2HeadersValidator` as it doesn't rely on the ordering

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            for (Entry<CharSequence, CharSequence> entry : headers) {
+                final CharSequence key = entry.getKey();
+                if (!hasPseudoHeaderFormat(key)) {
+                    // We know that pseudo header appears first so we can stop
+                    // looking once we get to the first non pseudo headers.
+                    break;
+                }
+
+                final PseudoHeaderName pseudoHeader = PseudoHeaderName.getPseudoHeader(key);
+                if (previousPseudoHeader != null &&
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Since this was moved out of HpackDecoder I think we should now validate the request pseudo headers as it leaves a hole otherwise.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                    pseudoHeader.isRequestOnly() != previousPseudoHeader.isRequestOnly()) {
+                    throw streamError(streamId, PROTOCOL_ERROR, "Mix of request and response pseudo-headers.");
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Since we're only validating the request specific headers I think we can make this another `else` clause in the branches below.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                }
+                previousPseudoHeader = pseudoHeader;
+
+                if (METHOD.value().contentEqualsIgnoreCase(key)) {
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: H2 headers must always have lower case keys and pseudo-headers should be no exception, so we should be able to omit the `IngoreCase`.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "TE header must not contain any value other than \"%s\"", TRAILERS);
+        }
+    }
+
+    /**
+     * Validates Pseudo-Header fields according to
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.3">RFC7540, section-8.1.2.3</a>
+     */
+    static void validatePseudoHeaders(Http2Headers headers, int streamId) throws Http2Exception {
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: Even if we don't plan on doing the client side now we should probably prepare for that by making the name reflect that it is request specific.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java [hunk 15]
```diff
@@ -571,6 +589,22 @@ private void verifyStreamMayHaveExisted(int streamId) throws Http2Exception {
...
+                (!stream.isHeadersReceived() || stream.state() == HALF_CLOSED_REMOTE)) {
+                validatePseudoHeaders(headers, streamId);
+            }
+
+            validateConnectionSpecificHeaders(headers, streamId);
+
+            if ((isInformational || !endOfStream) && stream.isHeadersReceived() || stream.isTrailersReceived()) {
```
[the remark this answers is not in the document]
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: Oh yeah good point, will wait on other opinion for now

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java (hunk 11 above)
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: This comment and the next one make me think we should probably leave this check in HPackDecoder.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (madgnome) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: I moved that out of the condition in the end.

**Comment by madgnome** [comment]
Any other opinion? @ejona86, @Scottmitch ?

**Comment by normanmaurer** [comment]
@ejona86 @carl-mastrangelo @bryce-anderson @Scottmitch @mosesn @rkapsi  so what we want to do with this ? Sorry I totally lost track of this one :/

**Comment by carl-mastrangelo** [comment]
@normanmaurer I like the idea of validation, but I don't know if I want to take on the performance penalty (Ctrl F for "bench" doesn't bring up anything).    In the case where most clients and servers are well behaved (i.e. both owned by the same party) it doesn't make sense to add so much validation.  Not sure this would catch any violations.

That said, if @madgnome can show that the default on case is not so expensive, following the spec more strictly can only be good for Netty users.

**Comment by normanmaurer** [comment]
@madgnome @carl-mastrangelo @ejona86 @bryce-anderson what do we want to do with this one ? I think I am fine with adding this as long as we use a default of false for now. WDYT ?

**Comment by bryce-anderson** [comment]
I'm fine with introducing it with a default of disabled.

**Comment by normanmaurer** [comment]
@carl-mastrangelo @ejona86 @trustin WDYT?

**Comment by trustin** [comment]
+1. Maybe better defaulting to 'enabled' and allowing to disable?

**Comment by normanmaurer** [comment]
@trustin I am a bit afraid that this will break apps when upgrading ... you are not ?

**Comment by trustin** [comment]
Shouldn't we consider this as a bug fix? What requests that violate the validation rules in this PR could be sent in reality?

**Comment by normanmaurer** [comment]
@trustin I think it is more about what @bryce-anderson mentioned at some point... It may hurt proxies.

**Comment by trustin** [comment]
What if we validate only pseudo headers and TE strictly and remove the connection headers from HTTP/1 then?

**Comment by normanmaurer** [comment]
@bryce-anderson @ejona86 @carl-mastrangelo WDYT ?

**Comment by bryce-anderson** [comment]
Will there be an option to disable all this new validation? I'm not strongly opinionated about it unless we can't disable it.

**Comment by normanmaurer** [comment]
Wouldn’t it be good enough to allow to disable this with all the other header validation?

**Comment by normanmaurer** [comment]
rebased this one

**Comment by normanmaurer** [comment]
@bryce-anderson sorry missed to tag you on my response...

**Comment by bryce-anderson** [comment]
@normanmaurer, yes, that should be fine if it can be disabled with the rest of the validation.

**Comment by normanmaurer** [comment]
Let me merge this ... As stated it can be disabled with the rest of the validation and follows more closely the RFC. We will add a note in the release note.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    /**
+     * Validates connection-specific headers according to
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.2">RFC7540, section-8.1.2.2</a>
+     */
+    static void validateConnectionSpecificHeaders(Http2Headers headers, int streamId) throws Http2Exception {
+        for (int i = 0; i < connectionSpecificHeaders.size(); i++) {
+            final AsciiString header = connectionSpecificHeaders.get(i);
+            if (headers.contains(header)) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Connection-speficic headers like [%s] must not be used with HTTP/2.", header);
```
Comment (jplock) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java]: spelling: specific

**Comment by normanmaurer** [comment]
@madgnome thanks a lot! Sorry for the long turn-around
